vm/vmvalidation.nim

```diff
@@ -0,0 +1,507 @@
+
+const
+  errSubroutineJump = "jump across subroutine boundaries"
+  errNotForwardJump = "not a forward jump"
+  errNotAProcType   = "type is not a proc type"
+
+proc toValueType(t: VmType): ValueType =
+  case t.kind
+  of tkInt, tkProc: vtInt
+  of tkFloat:       vtFloat
+  of tkForeign:     vtRef
+  of tkVoid:        unreachable()
+
+proc param(env: TypeEnv, t: TypeId, i: int): ValueType =
+  toValueType(env[env.params[env[t].a + i.uint32 + 1]])
+
+proc numArgs(typ: VmType): int =
+  typ.len - 1
+
+proc test(a: seq|openArray, i: SomeInteger): bool =
+  when i is SomeUnsignedInt:
+    i < a.len.uint
+  else:
+    i >= 0 and i < a.len
+
+template check(cond: bool, msg: untyped) =
+  ## Helper template for easier error propagation.
+  if not cond:
+    return CheckResult.err(msg)
+
+template check(res: CheckResult) =
+  ## Helper template for easier error propagation.
+  let tmp = res
+  if tmp.isErr:
+    return tmp
+
+proc firstPass(ctx: var ValidationState, code: openArray[Instr]): CheckResult =
+  ## * gathers and validates the existing subroutines
+  ## * makes sure jump targets are valid
+  ## * fills the `targets` set
+  template checkTarget(src: int, rel: int32) =
+    check src + rel in 0..(ctx.length-1), "jump target is not valid"
+
+  template jump(src: int, rel: int32) =
+    checkTarget(src, rel)
+    ctx.targets.incl PrgCtr(src + rel)
+
+  var stack: seq[int]
+
+  for pc, instr in code.pairs:
+    case instr.opcode
+    of opcJmp, opcBranch:
+      jump(pc, imm32(instr))
+    of opcEnter:
+      checkTarget(pc, imm32(instr))
+      # not a jump like the other instructions
+      check code[pc + imm32(instr)].opcode == opcBegin,
+            "target is not an 'opcBegin' instruction"
+    of opcLeave:
+      jump(pc.int32, imm32(instr))
+    of opcBegin:
+      stack.add ctx.subroutines.len
+      # the Begin instruction is not included in the covered range
+      ctx.subroutines.add Subroutine(covers: PrgCtr(pc + 1)..PrgCtr(0))
+    of opcEnd:
+      check stack.len > 0, "not in subroutine"
+      let pos = stack.pop()
+      ctx.subroutines[pos].covers.b = PrgCtr(pc)
+      ctx.subroutines[pos].parent =
+        if stack.len > 0: stack[^1]
+        else:             -1
+    else:
+      discard
+
+  check stack.len == 0, "unclosed subroutines exist"
+  result.initSuccess()
+
+proc find(ctx: ValidationState, pc: PrgCtr): int =
+  for i in countdown(ctx.subroutines.high, 0):
+    if pc in ctx.subroutines[i].covers:
+      return i
+  result = -1
+
+proc computeDepth(ctx: ValidationState, start: PrgCtr, target: PrgCtr): int =
+  ## Computes and returns the amount of subroutines a forward jump from `start`
+  ## to `target` exits. If a nested subroutine is entered - which is illegal -,
+  ## -1 is returned.
+  if ctx.subroutines.len == 0:
+    return 0
+
+  let
+    a = find(ctx, start)
+    b = find(ctx, target)
+
+  if a == b:
+    return 0 # intra-subroutine jump
+  elif a == -1:
+    return -1 # fail fast
+  else:
```

Comment:
Not quite, because both `result = -1` and `result = subroutines.len` would not always be correct.

Consider the following bytecode:
```
  Begin      # subroutine 0
  End
  Begin      # subroutine 1
  Leave L1 1 # <- 'a'
  End
.label L1
  Ret        # <- 'b'
```
When `computeDepth` is called to compute the number of subroutines the `Leave` instruction exits, `a` will be 1 (because the `Leave` instruction is part of subroutine 1) and `b` will be -1 (because the jump target - i.e., the `Ret` instruction - is not part of any subroutine). The parent subroutine of subroutine 1 is -1 (indicating that it has no parent), so the `while` loop will run once, and since `b` is -1, the `if` won't be entered and 1 (the depth) is returned.

The `a == -1` and `b >= 0` case can fail fast, since it means that there is a jump from outside a subroutine *into* one, and that's always illegal.

I'll improve the documentation for the code a bit, in order to make it more clear what is happening and why.